packages/babel-parser/src/util/expression-scope.js

```diff
@@ -0,0 +1,183 @@
+// @flow
+
+/*:: declare var invariant; */
+/**
+ * @module util/expression-scope
+
+ExpressionScope is used to track declaration errors in these ambiguous patterns:
+
+- CoverParenthesizedExpressionAndArrowParameterList
+  e.g. we don't know if `({ x })` is an parenthesized expression or an
+  arrow function parameters until we see an `=>` after `)`.
+
+- CoverCallExpressionAndAsyncArrowHead
+   e.g. we don't know if `async({ x })` is a call expression or an async arrow
+   function parameters until we see an `=>` after `)`
+
+The following declaration errors (@see parser/error-message) will be recorded in
+some expression scopes and thrown later when we know what the ambigous pattern is
+
+- AwaitBindingIdentifier
+- AwaitExpressionFormalParameter
+- YieldInParameter
+
+There are four different expression scope
+- Expression
+  A general scope that represents program / function body / static block. No errors
+  will be recorded nor thrown in this scope.
+
+- MaybeArrowParameterDeclaration
+  A scope that represents ambiguous arrow head e.g. `(x)`. Errors will be recorded
+  alongside parent scopes and thrown when `ExpressionScopeHandler#validateAsPattern`
+  is called.
+
+- MaybeAsyncArrowParameterDeclaration
+  A scope that represents ambiguous async arrow head e.g. `async(x)`. Errors will
+  be recorded alongside parent scopes and thrown when
+  `ExpressionScopeHandler#validateAsPattern` is called.
+
+- ParameterDeclaration
+  A scope that represents unambiguous function parameters `function(x)`. Errors
+  recorded in this scope will be thrown immediately. No errors will be recorded in
+  this scope.
+
+// @see {@link https://docs.google.com/document/d/1FAvEp9EUK-G8kHfDIEo_385Hs2SUBCYbJ5H-NnLvq8M|V8 Expression Scope design docs}
+ */
+
+const kExpression = 0,
+  kMaybeArrowParameterDeclaration = 1,
+  kMaybeAsyncArrowParameterDeclaration = 2,
+  kParameterDeclaration = 3;
+
+type ExpressionScopeType = 0 | 1 | 2 | 3;
+
+type raiseFunction = (number, string, ...any) => void;
+
+class ExpressionScope {
+  type: ExpressionScopeType;
+
+  constructor(type: ExpressionScopeType = kExpression) {
+    this.type = type;
+  }
+
+  canBeArrowParameterDeclaration() {
+    return (
+      this.type === kMaybeAsyncArrowParameterDeclaration ||
+      this.type === kMaybeArrowParameterDeclaration
+    );
+  }
+
+  isCertainlyParameterDeclaration() {
+    return this.type === kParameterDeclaration;
+  }
+}
+
+class ArrowHeadParsingScope extends ExpressionScope {
+  declare declarationErrorPos: number;
+  declare declarationErrorMessage: string;
```

Comment:
When in `errorRecovery` mode, there could be multiple errors. For example, in `async (x = await, y = await) => {}`.

We can thus store this as a map, and track all of them:
```suggestion
  errors: Map</* pos */ number, /* message */ string> = new Map();
```